docs/book/command-params.md

```diff
@@ -0,0 +1,378 @@
+# Command Params
+
+The Symfony Console component allows commands to define
+[input arguments and options](https://symfony.com/doc/current/console/input.html).
+
+The laminas-cli package adds a third possibility, **input parameters**, which
+allow you to define command options that will become interactive prompts when
+omitted during command invocation.
+
+## Using input parameters
+
+Internally, input parameters behave like standard options with the following
+modifications:
+
+- Input parameters are optional by default.
+- In interactive mode, when retrieving a parameter value that was not provided
+  as an option, the user will be asked to provide the value.
+- In non-interactive mode, if the user does not supply the option during
+  invocation, and the parameter is not marked as required, its
+  default value will be used; otherwise, an exception will be thrown,
+- **If the parameter has validations or normalizations supplied, these will be
+  applied regardless of whether the option is provided during invocation or when
+  the user is prompted to provide it.**
+- If the user provides a value via a prompt, that value will be stored as an
+  input option so it can be passed to the [next command in the
+  chain](commands-chain.md) (if one is defined).
+
+## Input parameter types
+
+Input parameters are classes that implement
+`Laminas\Cli\Input\InputParamInterface`:
+
+```php
+namespace Laminas\Cli\Input;
+
+use Symfony\Component\Console\Question\Question;
+
+interface InputParamInterface
+{
+    /**
+     * Default value to use if none provided.
+     *
+     * @return null|mixed
+     */
+    public function getDefault();
+
+    public function getDescription(): string;
+
+    public function getName(): string;
+
+    /**
+     * Return the Symfony\Component\Console\Input\InputOption::VALUE_* type.
+     */
+    public function getOptionMode(): ?int;
```

Comment:
Didn't we state somewhere, that the option is per default `optional` unless otherwise configured?
Would prefer having `int` as return type instead of nullable int.

Reply:
That's actually good point. Probably it shouldn't be nullable here. @weierophinney ?

Reply:
Okay, I'll change it to non-nullable `int`.

In terms of array options - that's up to the individual parameter implementations. We don't currently supply one, but theoretically, it's possible to create one; you'd just have to create a question that reprompts until the user has provided all the values they want to.